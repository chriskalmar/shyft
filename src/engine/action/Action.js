import { passOrThrow, isMap, isFunction } from '../util';

import {
  generatePermissionDescription,
  processActionPermissions,
} from '../permission/Permission';

export const ACTION_TYPE_MUTATION = 'mutation';
export const ACTION_TYPE_QUERY = 'query';
export const actionTypes = [ ACTION_TYPE_MUTATION, ACTION_TYPE_QUERY ];

class Action {
  constructor(setup = {}) {
    const {
      name,
      description,
      input,
      output,
      resolve,
      type,
      permissions,
    } = setup;

    passOrThrow(name, () => 'Missing action name');
    passOrThrow(description, () => `Missing description for action '${name}'`);

    passOrThrow(
      !input || isMap(input) || isFunction(input),
      () => `Action '${name}' has an invalid input definition`,
    );

    passOrThrow(
      !output || isMap(output) || isFunction(output),
      () => `Action '${name}' has an invalid output definition`,
    );

    passOrThrow(
      isFunction(resolve),
      () => `Action '${name}' needs a resolve function`,
    );

    passOrThrow(
      !type || actionTypes.indexOf(type) >= 0,
      () =>
        `Unknown action type '${type}' used in action '${name}', try one of these: '${actionTypes.join(
          ', ',
        )}'`,
    );

    this.name = name;
    this.description = description;
    this.input = input;
    this.output = output;
    this.resolve = resolve;
    this.type = type || ACTION_TYPE_MUTATION;
    this._permissions = permissions;
  }

  getInput() {
    if (!this.hasInput()) {
      return null;
    }

    if (this._input) {
      return this._input;
    }

    if (isFunction(this.input)) {
      this.input = this.input();

      passOrThrow(
        isMap(this.input),
        () =>
          `Input definition function for action '${
            this.name
          }' does not return a map`,
      );
    }

    passOrThrow(
      this.input.type,
      () => `Missing input type for action '${this.name}'`,
    );

    if (isFunction(this.input.type)) {
      this.input.type = this.input.type({
        name: 'input',
        description: this.input.description || this.description,
      });
    }

    this._input = this.input;

    return this._input;
  }

  hasInput() {
    return !!this.input;
  }

  getOutput() {
    if (!this.hasOutput()) {
      return null;
    }

    if (this._output) {
      return this._output;
    }

    if (isFunction(this.output)) {
      this.output = this.output();

      passOrThrow(
        isMap(this.output),
        () =>
          `Output definition function for action '${
            this.name
          }' does not return a map`,
      );
    }

    passOrThrow(
      this.output.type,
      () => `Missing output type for action '${this.name}'`,
    );

    if (isFunction(this.output.type)) {
      this.output.type = this.output.type({
        name: 'output',
        description: this.output.description || this.description,
      });
    }

    this._output = this.output;

    return this._output;
  }

  hasOutput() {
    return !!this.output;
  }

  _processPermissions() {
    if (this._permissions) {
      const permissions = isFunction(this._permissions)
        ? this._permissions()
        : this._permissions;

      return processActionPermissions(this, permissions);
    }
    else if (this._defaultPermissions) {
      return processActionPermissions(this, this._defaultPermissions);
    }

    return null;
  }

  _generatePermissionDescriptions() {
    if (this.permissions) {
      this.descriptionPermissions = generatePermissionDescription(
        this.permissions,
      );
    }
  }

  _injectDefaultPermissionsBySchema(defaultPermissions) {
    this._defaultPermissions = defaultPermissions;
  }

  getPermissions() {
    if ((!this._permissions && !this._defaultPermissions) || this.permissions) {
      return this.permissions;
    }

    this.permissions = this._processPermissions();
    this._generatePermissionDescriptions();
    return this.permissions;
  }

  toString() {
    return this.name;
  }
}

export default Action;

export const isAction = obj => {
  return obj instanceof Action;
};
