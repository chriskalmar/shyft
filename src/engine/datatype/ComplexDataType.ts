export class ComplexDataType {}

export const isComplexDataType = (obj: unknown): obj is ComplexDataType => {
  return obj instanceof ComplexDataType;
};
