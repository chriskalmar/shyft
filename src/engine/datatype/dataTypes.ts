import { DataType } from './DataType';
import { DataTypeUser } from './DataTypeUser';
import { randomJson } from '../util';
import * as casual from 'casual';

export const DataTypeUserID = new DataTypeUser({
  name: 'DataTypeUserID',
  description: 'Data type representing a reference to a user',
  /* istanbul ignore next */
  mock: () => casual.integer(2 ^ 20, 2 ^ 51).toString(),
});

export const DataTypeID = new DataType({
  name: 'DataTypeID',
  description: 'Data type representing unique IDs',
  /* istanbul ignore next */
  mock: () => casual.integer(2 ^ 20, 2 ^ 51).toString(),
});

export const DataTypeInteger = new DataType({
  name: 'DataTypeInteger',
  description: 'Data type representing integer values',
  /* istanbul ignore next */
  mock: () => casual.integer(-2 ^ 10, 2 ^ 10),
});

export const DataTypeBigInt = new DataType({
  name: 'DataTypeBigInt',
  description: 'Data type representing big integer values',
  /* istanbul ignore next */
  mock: () => casual.integer(2 ^ 20, 2 ^ 51).toString(),
});

export const DataTypeFloat = new DataType({
  name: 'DataTypeFloat',
  description: 'Data type representing float values',
  /* istanbul ignore next */
  mock: () => casual.double(-2 ^ 10, 2 ^ 10),
});

export const DataTypeDouble = new DataType({
  name: 'DataTypeDouble',
  description: 'Data type representing double precision values',
  /* istanbul ignore next */
  mock: () => casual.double(-2 ^ 10, 2 ^ 10),
});

export const DataTypeBoolean = new DataType({
  name: 'DataTypeBoolean',
  description: 'Data type representing boolean values',
  /* istanbul ignore next */
  mock: () => casual.boolean,
  enforceRequired: true,
  defaultValue() {
    return false;
  },
});

export const DataTypeString = new DataType({
  name: 'DataTypeString',
  description: 'Data type representing text values',
  /* istanbul ignore next */
  mock: () => casual.title,
});

export const DataTypeJson = new DataType({
  name: 'DataTypeJson',
  description: 'Data type representing json objects',
  /* istanbul ignore next */
  mock: randomJson,
});

export const DataTypeTimestamp = new DataType({
  name: 'DataTypeTimestamp',
  description: 'Data type representing a timestamp',
  /* istanbul ignore next */
  mock: () => new Date(casual.unix_time * 1000),
});

export const DataTypeTimestampTz = new DataType({
  name: 'DataTypeTimestampTz',
  description: 'Data type representing a timestamp with time zone information',
  /* istanbul ignore next */
  mock: () => new Date(casual.unix_time * 1000),
});

export const DataTypeDate = new DataType({
  name: 'DataTypeDate',
  description: 'Data type representing a date',
  /* istanbul ignore next */
  mock: () => new Date(casual.unix_time * 1000),
});

export const DataTypeTime = new DataType({
  name: 'DataTypeTime',
  description: 'Data type representing time',
  /* istanbul ignore next */
  mock: () => new Date(casual.unix_time * 1000),
});

export const DataTypeTimeTz = new DataType({
  name: 'DataTypeTimeTz',
  description: 'Data type representing time with time zone information',
  /* istanbul ignore next */
  mock: () => new Date(casual.unix_time * 1000),
});

export const DataTypeUUID = new DataType({
  name: 'DataTypeUUID',
  description: 'Data type representing a UUID',
  /* istanbul ignore next */
  mock: () => casual.uuid,
});

export const DataTypeI18n = new DataType({
  name: 'DataTypeI18n',
  description: 'Data type representing translation objects',
  /* istanbul ignore next */
  mock: randomJson,
});

export const DataTypeUpload = new DataType({
  name: 'DataTypeUpload',
  description: 'Data type representing a file upload',
  mock: () => null,
});
