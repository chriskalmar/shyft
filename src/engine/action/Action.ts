/* eslint-disable @typescript-eslint/no-explicit-any */

import { passOrThrow, isMap, isFunction } from '../util';
import { AttributeBase } from '../attribute/Attribute';
import { DataTypeFunction } from '../datatype/DataType';
import {
  generatePermissionDescription,
  processActionPermissions,
  Permission,
} from '../permission/Permission';
import { Context } from '../context/Context';
import { GraphQLResolveInfo, Source } from 'graphql';

export const ACTION_TYPE_MUTATION = 'mutation';
export const ACTION_TYPE_QUERY = 'query';
export const actionTypes = [ACTION_TYPE_MUTATION, ACTION_TYPE_QUERY];

export type ActionPreProcessor = (params: {
  action: Action;
  source?: any;
  input?: { [key: string]: unknown };
  context?: Context;
  info?: GraphQLResolveInfo;
}) => void | Promise<void>;

export type ActionPostProcessor = (params: {
  result?: { [key: string]: unknown };
  error?: Error;
  action: Action;
  source?: Source;
  input?: { [key: string]: unknown };
  context?: Context;
  info?: GraphQLResolveInfo;
}) => void | Promise<void>;

export type ActionResolver = (params: {
  source?: Source;
  input?: { [key: string]: unknown };
  context?: Context;
  info?: GraphQLResolveInfo;
}) => unknown | Promise<unknown>;

export type ActionSetup = {
  name?: string;
  description?: string;
  // input?: AttributeBase | Function | { [type: string]: Function };
  input?: any;
  // output?: AttributeBase | Function | { [type: string]: Function };
  output?: any;
  resolve: ActionResolver;
  type?: string;
  permissions?: ((...args) => Permission | Permission[]) | Permission | Permission[];
  // preProcessor?: Function;
  // postProcessor?: Function;
  preProcessor?: ActionPreProcessor;
  postProcessor?: ActionPostProcessor;
};

export class Action {
  name: string;
  description: string;
  // input: AttributeBase | Function | { [type: string]: Function };
  // private _input: AttributeBase | Function | { [type: string]: Function };
  input: any;
  private _input: any;
  // output: AttributeBase | Function | { [type: string]: Function };
  // private _output: AttributeBase | Function | { [type: string]: Function };
  output: any;
  private _output: any;
  resolve: ActionResolver;
  type: string;
  permissions: Permission | Permission[];
  private readonly _permissions: ((...args) => Permission | Permission[]) | Permission | Permission[];
  private _defaultPermissions: Permission | Permission[];
  descriptionPermissions: string | boolean;
  preProcessor?: ActionPreProcessor;
  postProcessor?: ActionPostProcessor;

  constructor(setup: ActionSetup = {} as ActionSetup) {
    const {
      name,
      description,
      input,
      output,
      resolve,
      type,
      permissions,
      preProcessor,
      postProcessor,
    } = setup;

    passOrThrow(name, () => 'Missing action name');
    passOrThrow(description, () => `Missing description for action '${name}'`);

    passOrThrow(
      !input || isMap(input) || isFunction(input),
      () => `Action '${name}' has an invalid input definition`,
    );

    passOrThrow(
      !output || isMap(output) || isFunction(output),
      () => `Action '${name}' has an invalid output definition`,
    );

    passOrThrow(
      isFunction(resolve),
      () => `Action '${name}' needs a resolve function`,
    );

    passOrThrow(
      !type || actionTypes.indexOf(type) >= 0,
      () =>
        `Unknown action type '${type}' used in action '${name}', try one of these: '${actionTypes.join(
          ', ',
        )}'`,
    );

    if (preProcessor) {
      passOrThrow(
        isFunction(preProcessor),
        () => `preProcessor of action '${name}' needs to be a valid function`,
      );

      this.preProcessor = preProcessor;
    }

    if (postProcessor) {
      passOrThrow(
        isFunction(postProcessor),
        () => `postProcessor of action '${name}' needs to be a valid function`,
      );

      this.postProcessor = postProcessor;
    }

    this.name = name;
    this.description = description;
    this.input = input;
    this.output = output;
    this.resolve = resolve;
    this.type = type || ACTION_TYPE_MUTATION;
    this._permissions = permissions;
  }

  getInput(): null | any {
    if (!this.hasInput()) {
      return null;
    }

    if (this._input) {
      return this._input;
    }

    if (isFunction(this.input)) {
      const inputFn = this.input as (...args) => any;
      this.input = inputFn();

      passOrThrow(
        isMap(this.input),
        () =>
          `Input definition function for action '${this.name}' does not return a map`,
      );
    }

    const inputAttr = this.input as AttributeBase;
    passOrThrow(
      inputAttr.type,
      () => `Missing input type for action '${this.name}'`,
    );

    if (isFunction(inputAttr.type)) {
      const inputAttrType = inputAttr.type as DataTypeFunction;
      this.input = this.input as AttributeBase;
      this.input.type = inputAttrType({
        setup: {
          name: 'input',
          description: this.input.description || this.description,
        },
      });
    }

    this._input = this.input;

    return this._input;
  }

  hasInput(): boolean {
    return !!this.input;
  }

  getOutput(): null | any {
    if (!this.hasOutput()) {
      return null;
    }

    if (this._output) {
      return this._output;
    }

    if (isFunction(this.output)) {
      const outputFn = this.output as (...args) => any;
      this.output = outputFn();

      passOrThrow(
        isMap(this.output),
        () =>
          `Output definition function for action '${this.name}' does not return a map`,
      );
    }

    const outputAttr = this.output as AttributeBase;
    passOrThrow(
      outputAttr.type,
      () => `Missing output type for action '${this.name}'`,
    );

    if (isFunction(outputAttr.type)) {
      const outputAttrType = outputAttr.type as DataTypeFunction;
      this.output = this.output as AttributeBase;
      this.output.type = outputAttrType({
        setup: {
          name: 'output',
          description: this.output.description || this.description,
        },
      });
    }

    this._output = this.output;

    return this._output;
  }

  hasOutput(): boolean {
    return !!this.output;
  }

  _processPermissions(): null | Permission | Permission[] {
    if (this._permissions) {
      if (isFunction(this._permissions)) {
        const permissionsFn = this._permissions as (...args) => Permission | Permission[];
        const permissions = permissionsFn();
        return processActionPermissions(this, permissions);
      }
      const permissions = this._permissions as Permission | Permission[];
      return processActionPermissions(this, permissions);
    } else if (this._defaultPermissions) {
      return processActionPermissions(this, this._defaultPermissions);
    }

    return null;
  }

  _generatePermissionDescriptions(): void {
    if (this.permissions) {
      let permissions: Permission | Permission[];
      if (isFunction(this._permissions)) {
        const permissionsFn = this._permissions as (...args) => Permission | Permission[];
        permissions = permissionsFn();
      } else {
        permissions = this._permissions as Permission | Permission[];
      }

      this.descriptionPermissions = generatePermissionDescription(permissions);
    }
  }

  _injectDefaultPermissionsBySchema(defaultPermissions): void {
    this._defaultPermissions = defaultPermissions;
  }

  getPermissions(): ((...args) => Permission | Permission[]) | Permission | Permission[] {
    if ((!this._permissions && !this._defaultPermissions) || this.permissions) {
      return this.permissions;
    }

    this.permissions = this._processPermissions();
    this._generatePermissionDescriptions();
    return this.permissions;
  }

  toString(): string {
    return this.name;
  }
}

export const isAction = (obj: unknown): obj is Action => {
  return obj instanceof Action;
};
