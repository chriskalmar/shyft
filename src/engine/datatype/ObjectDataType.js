import { passOrThrow, resolveFunctionMap, isMap, isFunction } from '../util';
import { isEntity } from '../entity/Entity';
import { isDataType } from './DataType';
import { ComplexDataType, isComplexDataType } from './ComplexDataType';

export class ObjectDataType extends ComplexDataType {
  constructor(setup = {}) {
    super();

    const { name, description, attributes } = setup;

    passOrThrow(name, () => 'Missing object data type name');
    passOrThrow(
      description,
      () => `Missing description for object data type '${name}'`,
    );
    passOrThrow(
      attributes,
      () => `Missing attributes for object data type '${name}'`,
    );

    passOrThrow(
      isMap(attributes) || isFunction(attributes),
      () =>
        `Object data type '${name}' needs an attribute definition as a map or a function returning such a map`,
    );

    this.name = name;
    this.description = description;
    this._attributesMap = attributes;
  }

  getAttributes() {
    if (this._attributes) {
      return this._attributes;
    }

    const ret = (this._attributes = this._processAttributeMap());
    return ret;
  }

  _processAttribute(rawAttribute, attributeName) {
    if (isFunction(rawAttribute.type)) {
      rawAttribute.type = rawAttribute.type({
        name: attributeName,
        description: rawAttribute.description,
      });
    }

    const attribute = {
      ...rawAttribute,
      required: !!rawAttribute.required,
      name: attributeName,
    };

    passOrThrow(
      attribute.description,
      () => `Missing description for '${this.name}.${attributeName}'`,
    );

    passOrThrow(
      isDataType(attribute.type) ||
        isComplexDataType(attribute.type) ||
        isEntity(attribute.type),
      () =>
        `'${this.name}.${attributeName}' has invalid data type '${String(
          attribute.type,
        )}'`,
    );

    passOrThrow(
      !attribute.resolve || isFunction(attribute.resolve),
      () => `'${this.name}.${attributeName}' has an invalid resolve function'`,
    );

    if (attribute.defaultValue) {
      // enforce mandatory param if defaultValue provided
      attribute.required = true;

      passOrThrow(
        isFunction(attribute.defaultValue),
        () =>
          `'${
            this.name
          }.${attributeName}' has an invalid defaultValue function'`,
      );

      passOrThrow(
        !isComplexDataType(attribute.type),
        () =>
          `Complex data type '${
            this.name
          }.${attributeName}' cannot have a defaultValue function'`,
      );
    }

    passOrThrow(
      !attribute.validate || isFunction(attribute.validate),
      () => `'${this.name}.${attributeName}' has an invalid validate function'`,
    );

    return attribute;
  }

  _processAttributeMap() {
    // if it's a function, resolve it to get that map
    const attributeMap = resolveFunctionMap(this._attributesMap);

    passOrThrow(
      isMap(attributeMap),
      () =>
        `Attribute definition function for object data type '${
          this.name
        }' does not return a map`,
    );

    const attributeNames = Object.keys(attributeMap);
    passOrThrow(
      attributeNames.length > 0,
      () => `Object data type '${this.name}' has no attributes defined`,
    );

    const resultAttributes = {};

    attributeNames.forEach(attributeName => {
      resultAttributes[attributeName] = this._processAttribute(
        attributeMap[attributeName],
        attributeName,
      );
    });

    return resultAttributes;
  }

  validate = value => {
    if (value) {
      passOrThrow(
        isMap(value),
        () => `Object data type '${this.name}' expects an object`,
      );
    }
  };

  toString() {
    return this.name;
  }
}

export const isObjectDataType = obj => {
  return obj instanceof ObjectDataType;
};

export const buildObjectDataType = obj => {
  return ({ name, description }) =>
    new ObjectDataType({
      description,
      ...obj,
      name,
    });
};
