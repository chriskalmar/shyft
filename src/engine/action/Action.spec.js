import Action, { isAction } from './Action';
import Permission, { isPermission } from '../permission/Permission';

import { DataTypeString } from '../datatype/dataTypes';
import { buildObjectDataType } from '../datatype/ObjectDataType';

import { passOrThrow } from '../util';

describe('Action', () => {
  it('should have a name', () => {
    function fn() {
      // eslint-disable-next-line no-new
      new Action();
    }

    expect(fn).toThrowErrorMatchingSnapshot();
  });

  it('should have a description', () => {
    function fn() {
      // eslint-disable-next-line no-new
      new Action({
        name: 'example',
      });
    }

    expect(fn).toThrowErrorMatchingSnapshot();
  });

  it('should allow definitions without input', () => {
    function fn() {
      // eslint-disable-next-line no-new
      new Action({
        name: 'example',
        description: 'do something',
        resolve() {},
      });
    }

    expect(fn).not.toThrow();
  });

  it('should have a valid input definition', () => {
    function fn() {
      // eslint-disable-next-line no-new
      new Action({
        name: 'example',
        description: 'do something',
        input: 123,
        output: 456,
      });
    }

    expect(fn).toThrowErrorMatchingSnapshot();
  });

  it('should allow definitions without output', () => {
    function fn() {
      // eslint-disable-next-line no-new
      new Action({
        name: 'example',
        description: 'do something',
        input() {},
        resolve() {},
      });
    }

    expect(fn).not.toThrow();
  });

  it('should have a valid output definition', () => {
    function fn() {
      // eslint-disable-next-line no-new
      new Action({
        name: 'example',
        description: 'do something',
        input() {},
        output: 'wrong',
      });
    }

    expect(fn).toThrowErrorMatchingSnapshot();
  });

  it('should have valid input and output definitions when generated via a function', () => {
    // eslint-disable-next-line no-new
    const action = new Action({
      name: 'example',
      description: 'do something',
      input() {
        return 123;
      },
      output() {},
      resolve() {},
    });

    function fn1() {
      action.getInput();
    }

    function fn2() {
      action.getOutput();
    }

    expect(fn1).toThrowErrorMatchingSnapshot();
    expect(fn2).toThrowErrorMatchingSnapshot();
  });

  it('should have valid input data types', () => {
    const action = new Action({
      name: 'example',
      description: 'do something',
      input: {
        someAttr: {},
      },
      output: {},
      resolve() {},
    });

    function fn() {
      action.getInput();
    }

    expect(fn).toThrowErrorMatchingSnapshot();
  });

  it('should have valid output data types', () => {
    const action = new Action({
      name: 'example',
      description: 'do something',
      input: {},
      output: {
        someAttr: {},
      },
      resolve() {},
    });

    function fn() {
      action.getOutput();
    }

    expect(fn).toThrowErrorMatchingSnapshot();
  });

  // it('should have valid input attributes if valid defaultValue functions', () => {

  //   const action = new Action({
  //     name: 'example',
  //     description: 'do something',
  //     input: {
  //       type: DataTypeString,
  //       defaultValue: 'not-a-function'
  //     },
  //     output: {},
  //     resolve() {},
  //   })

  //   function fn() {
  //     action.getInput()
  //   }

  //   expect(fn).toThrowErrorMatchingSnapshot();

  // })

  it('should have a valid resolve function', () => {
    function fn() {
      // eslint-disable-next-line no-new
      new Action({
        name: 'example',
        description: 'do something',
      });
    }

    expect(fn).toThrowErrorMatchingSnapshot();
  });

  it("should return it's name", () => {
    const action = new Action({
      name: 'example',
      description: 'do something',
      input: {},
      output: {},
      resolve() {},
    });

    expect(action.name).toBe('example');
    expect(String(action)).toBe('example');
  });

  it('should return input and output params', () => {
    const action1 = new Action({
      name: 'example',
      description: 'do something',
      resolve() {},
    });

    const input1 = action1.getInput();
    const output1 = action1.getOutput();

    expect(input1).toEqual(null);
    expect(output1).toEqual(null);

    const hasInput1 = action1.hasInput();
    const hasOutput1 = action1.hasOutput();

    expect(hasInput1).toBe(false);
    expect(hasOutput1).toBe(false);

    const action2 = new Action({
      name: 'example',
      description: 'do something',
      input: {
        type: buildObjectDataType({
          attributes: {
            firstName: {
              description: 'First name',
              type: DataTypeString,
              defaultValue() {
                return 'Waldo';
              },
            },
            lastName: {
              description: 'Last name',
              type: DataTypeString,
              required: true,
            },
            about: {
              description: 'Just some description',
              type: buildObjectDataType({
                attributes: {
                  favouriteActor: {
                    type: DataTypeString,
                    description: 'One more description',
                  },
                },
              }),
            },
          },
        }),
      },
      output: {
        type: buildObjectDataType({
          attributes: {
            luckyNumber: {
              description: 'Lucky number',
              type: DataTypeString,
            },
          },
        }),
      },
      resolve() {},
    });

    const input2 = action2.getInput();
    const output2 = action2.getOutput();

    expect(input2).toMatchSnapshot();
    expect(output2).toMatchSnapshot();

    const attributes2 = input2.type.getAttributes();
    expect(attributes2).toMatchSnapshot();

    expect(typeof attributes2.firstName.defaultValue).toBe('function');

    const nestedAttributes2 = attributes2.about.type.getAttributes();
    expect(nestedAttributes2).toMatchSnapshot();

    const hasInput2 = action2.hasInput();
    const hasOutput2 = action2.hasOutput();

    expect(hasInput2).toBe(true);
    expect(hasOutput2).toBe(true);
  });

  describe('isAction', () => {
    const action = new Action({
      name: 'example',
      description: 'do something',
      input: {},
      output: {},
      resolve() {},
    });

    it('should recognize objects of type Action', () => {
      function fn() {
        passOrThrow(isAction(action), () => 'This error will never happen');
      }

      expect(fn).not.toThrow();
    });

    it('should recognize non-Action objects', () => {
      function fn() {
        passOrThrow(
          isAction({}) || isAction(function test() {}) || isAction(Error),
          () => 'Not a Action object',
        );
      }

      expect(fn).toThrowErrorMatchingSnapshot();
    });
  });

  describe('permissions', () => {
    it('should return permissions', () => {
      const action = new Action({
        name: 'example',
        description: 'do something',
        input: {},
        output: {},
        resolve() {},
        permissions: new Permission().authenticated(),
      });

      const permissions = action.getPermissions();

      expect(isPermission(permissions)).toBe(true);
    });

    it('should throw if empty permissions are provided', () => {
      const action = new Action({
        name: 'example',
        description: 'do something',
        input: {},
        output: {},
        resolve() {},
        permissions: [ new Permission().authenticated(), new Permission() ],
      });

      function fn() {
        action.getPermissions();
      }

      expect(fn).toThrowErrorMatchingSnapshot();
    });
  });
});
