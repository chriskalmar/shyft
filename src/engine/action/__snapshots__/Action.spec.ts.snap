// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Action defaultValue should fill in provided default value 1`] = `
Object {
  "data": Object {
    "someAction": Object {
      "result": Object {
        "value": 2000,
      },
    },
  },
}
`;

exports[`Action defaultValue should fill in provided default values in nested input objects 1`] = `
Object {
  "data": Object {
    "someAction": Object {
      "result": Object {
        "value": 2000,
      },
    },
  },
}
`;

exports[`Action isAction should recognize non-Action objects 1`] = `"Not a Action object"`;

exports[`Action permissions should throw if empty permissions are provided 1`] = `"Action 'example' has one or more empty permission definitions"`;

exports[`Action postProcessor should have a valid postProcessor function if defined 1`] = `"postProcessor of action 'example' needs to be a valid function"`;

exports[`Action postProcessor should pass through postProcessor if it is declared 1`] = `
Object {
  "data": Object {
    "someActionWithPostProcessor": Object {
      "result": Object {
        "value": 123,
      },
    },
  },
}
`;

exports[`Action postProcessor should pass through postProcessor if it is declared 2`] = `
Object {
  "data": Object {
    "someActionWithPostProcessor": Object {
      "result": Object {
        "value": 2468,
      },
    },
  },
}
`;

exports[`Action preProcessor should have a valid preProcessor function if defined 1`] = `"preProcessor of of action 'example' needs to be a valid function"`;

exports[`Action preProcessor should pass through preProcessor if it is declared 1`] = `
Object {
  "data": Object {
    "someActionWithPreProcessor": Object {
      "result": Object {
        "value": 123,
      },
    },
  },
}
`;

exports[`Action preProcessor should pass through preProcessor if it is declared 2`] = `
Object {
  "data": Object {
    "someActionWithPreProcessor": null,
  },
  "errors": Array [
    [GraphQLError: 13 brings bad luck],
  ],
}
`;

exports[`Action should have a description 1`] = `"Missing description for action 'example'"`;

exports[`Action should have a name 1`] = `"Missing action name"`;

exports[`Action should have a valid input definition 1`] = `"Action 'example' has an invalid input definition"`;

exports[`Action should have a valid output definition 1`] = `"Action 'example' has an invalid output definition"`;

exports[`Action should have a valid resolve function 1`] = `"Action 'example' needs a resolve function"`;

exports[`Action should have valid input and output definitions when generated via a function 1`] = `"Input definition function for action 'example' does not return a map"`;

exports[`Action should have valid input and output definitions when generated via a function 2`] = `"Output definition function for action 'example' does not return a map"`;

exports[`Action should have valid input data types 1`] = `"Missing input type for action 'example'"`;

exports[`Action should have valid output data types 1`] = `"Missing output type for action 'example'"`;

exports[`Action should return input and output params 1`] = `
Object {
  "type": ObjectDataType {
    "_attributesMap": Object {
      "about": Object {
        "description": "Just some description",
        "type": [Function],
      },
      "firstName": Object {
        "defaultValue": [Function],
        "description": "First name",
        "type": DataType {
          "description": "Data type representing text values",
          "mock": [Function],
          "name": "DataTypeString",
          "validate": [Function],
        },
      },
      "lastName": Object {
        "description": "Last name",
        "required": true,
        "type": DataType {
          "description": "Data type representing text values",
          "mock": [Function],
          "name": "DataTypeString",
          "validate": [Function],
        },
      },
    },
    "description": "do something",
    "name": "input",
    "validate": [Function],
  },
}
`;

exports[`Action should return input and output params 2`] = `
Object {
  "type": ObjectDataType {
    "_attributesMap": Object {
      "luckyNumber": Object {
        "description": "Lucky number",
        "type": DataType {
          "description": "Data type representing text values",
          "mock": [Function],
          "name": "DataTypeString",
          "validate": [Function],
        },
      },
    },
    "description": "do something",
    "name": "output",
    "validate": [Function],
  },
}
`;

exports[`Action should return input and output params 3`] = `
Object {
  "about": Object {
    "description": "Just some description",
    "name": "about",
    "required": false,
    "type": ObjectDataType {
      "_attributesMap": Object {
        "favouriteActor": Object {
          "description": "One more description",
          "type": DataType {
            "description": "Data type representing text values",
            "mock": [Function],
            "name": "DataTypeString",
            "validate": [Function],
          },
        },
      },
      "description": "Just some description",
      "name": "about",
      "validate": [Function],
    },
  },
  "firstName": Object {
    "defaultValue": [Function],
    "description": "First name",
    "name": "firstName",
    "required": true,
    "type": DataType {
      "description": "Data type representing text values",
      "mock": [Function],
      "name": "DataTypeString",
      "validate": [Function],
    },
  },
  "lastName": Object {
    "description": "Last name",
    "name": "lastName",
    "required": true,
    "type": DataType {
      "description": "Data type representing text values",
      "mock": [Function],
      "name": "DataTypeString",
      "validate": [Function],
    },
  },
}
`;

exports[`Action should return input and output params 4`] = `
Object {
  "favouriteActor": Object {
    "description": "One more description",
    "name": "favouriteActor",
    "required": false,
    "type": DataType {
      "description": "Data type representing text values",
      "mock": [Function],
      "name": "DataTypeString",
      "validate": [Function],
    },
  },
}
`;
